### PR TITLE
[quant][pt2] Preserve source_fn_stack after QAT fusion

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110899

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50101253](https://our.internmc.facebook.com/intern/diff/D50101253)

### CODE REVIEW

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 1]
```diff
@@ -528,6 +528,58 @@ def forward(self, x):
...
+        quantizer = XNNPACKQuantizer()
+        quantizer.set_global(get_symmetric_quantization_config(is_qat=True))
+        m = capture_pre_autograd_graph(m, example_inputs)
+        m = prepare_qat_pt2e(m, quantizer)
+        m(*example_inputs)
+        m = convert_pt2e(m)
+
+        # Assert these nodes have the right source partitions
+        node_name_to_source_partition = {
+            "_param_constant0": "l__self___backbone_conv",
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:555]: should we match the type string instead? that feels more stable than the name

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 2]
```diff
@@ -528,6 +528,58 @@ def forward(self, x):
...
+            "_param_constant1": "l__self___backbone_conv",
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:556]: also node_names are free to change, so we should not rely on it I think

File: torch/ao/quantization/pt2e/qat_utils.py [hunk 3]
```diff
@@ -571,8 +574,13 @@ def _fuse_conv_bn_qat_helper(m: GraphModule, is_cuda: bool) -> GraphModule:
...
     for r in replacements_with_conv_bias + replacements_no_conv_bias:
-        (replacement_conv_node, replacement_bn_node, replacement_getitem_node) =\
-            _get_conv_bn_getitem_nodes(r.replacements)
+        # TODO: rewrite this to return a map
+        (
+            replacement_conv_node,
+            replacement_bn_node,
+            replacement_getitem_node,
+            replacement_relu_node,
+        ) = _get_conv_bn_getitem_nodes(r.replacements)
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/qat_utils.py:583]: I guess this is for identifying the nodes, I feel https://github.com/pytorch/pytorch/pull/110743 should help

**Review by jerryzh168 — APPROVED** [review]
maybe we can extend https://github.com/pytorch/pytorch/pull/110743 to work with subgraph rewriter and query node by name

File: test/quantization/pt2e/test_quantize_pt2e_qat.py (hunk 2 above)
Comment (andrewor14) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:556]: sounds good, let me see what I can do

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

File: torch/ao/quantization/pt2e/qat_utils.py [hunk 4]
```diff
@@ -359,14 +363,14 @@ def _get_fused_convbn_q_dq_nodes(nodes: List[Node]) -> Tuple[Node, Node]:
     assert dq_node is not None
     return (q_node, dq_node)
 
-def _get_conv_bn_getitem_nodes(nodes: List[Node]) -> Tuple[Node, Node, Node]:
+def _get_conv_bn_getitem_nodes(nodes: List[Node]) -> Tuple[Node, Node, Node, Optional[Node]]:
```
Comment (kimishpatel) [inline torch/ao/quantization/pt2e/qat_utils.py:366]: should also be renamed to _get_conv_bn_getitem_relu_nodes

File: torch/ao/quantization/pt2e/qat_utils.py [hunk 5]
```diff
@@ -694,9 +710,29 @@ def _fold_conv_bn_qat_helper(m: GraphModule, is_cuda: bool) -> GraphModule:
...
+    # Step (2): Copy over metadata from original subgraph
+    for r in replacements:
+        (
+            replacement_conv_node,
+            replacement_bn_node,
+            replacement_getitem_node,
+            replacement_relu_node,
+        ) = _get_conv_bn_getitem_nodes(r.replacements)
+
+        for original_node in _filter_nodes_map(r.nodes_map).values():
```
Comment (kimishpatel) [inline torch/ao/quantization/pt2e/qat_utils.py:722]: Andrew, why do you need to copy the metadata?

Comment (andrewor14) [reply torch/ao/quantization/pt2e/qat_utils.py:722]: This is for lowering models trained by QAT. We found that the `source_fn` was not preserved and all convs in the model ended up being in the same partition `conv2d`, but they should be in different partitions

**Pull request closed without merging**